genie-web/src/main/java/com/netflix/genie/web/services/impl/S3AttachmentService.java

```diff
@@ -0,0 +1,225 @@
+import com.netflix.genie.web.properties.AttachmentServiceProperties;
+import com.netflix.genie.web.services.AttachmentService;
+import com.netflix.genie.web.util.MetricsUtils;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.core.io.Resource;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implementation of the AttachmentService interface which saves attachments to AWS S3.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class S3AttachmentService implements AttachmentService {
+
+    private static final String METRICS_PREFIX = "genie.jobs.attachments.s3";
+    private static final String COUNT_DISTRIBUTION = METRICS_PREFIX + ".count.distribution";
+    private static final String LARGEST_SIZE_DISTRIBUTION = METRICS_PREFIX + ".largest.distribution";
+    private static final String TOTAL_SIZE_DISTRIBUTION = METRICS_PREFIX + ".totalSize.distribution";
+    private static final String SAVE_TIMER = METRICS_PREFIX + ".upload.timer";
+    private static final Set<URI> EMPTY_SET = ImmutableSet.of();
+    private static final String SLASH = "/";
+    private static final String S3 = "s3";
+    private final S3ClientFactory s3ClientFactory;
+    private final AttachmentServiceProperties properties;
+    private final MeterRegistry meterRegistry;
+    private final AmazonS3URI s3BaseURI;
+
+    /**
+     * Constructor.
+     *
+     * @param s3ClientFactory             the s3 client factory
+     * @param attachmentServiceProperties the service properties
+     * @param meterRegistry               the meter registry
+     */
+    public S3AttachmentService(
+        final S3ClientFactory s3ClientFactory,
+        final AttachmentServiceProperties attachmentServiceProperties,
+        final MeterRegistry meterRegistry
+    ) {
+        this.s3ClientFactory = s3ClientFactory;
+        this.properties = attachmentServiceProperties;
+        this.meterRegistry = meterRegistry;
+        this.s3BaseURI = new AmazonS3URI(attachmentServiceProperties.getLocationPrefix());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<URI> saveAttachments(
+        @Nullable final String jobId,
+        final Set<Resource> attachments
+    ) throws SaveAttachmentException {
+
+        // Track number of attachments, including zeroes
+        this.meterRegistry.summary(COUNT_DISTRIBUTION).record(attachments.size());
+
+        log.debug("Saving {} attachments for job request with id: {}", attachments.size(), jobId);
+
+        if (attachments.size() == 0) {
+            return EMPTY_SET;
+        }
+
+        // Check for attachment size limits
+        this.checkLimits(attachments);
```

Comment:
shouldn't this all be within the try? as it is you've missed these in some counter metrics below since you return before they're logged

Reply:
That's intentional. The upload timer is only recorded if we reach the upload part.
And the distribution counters are only updated if the attachments are not rejected due to their size.

But maybe i should add a metric for limit exceeded.
(i didn't because it's easy to see with other existing metrics)